### ISSUE TITLE
discussion: Common ledger-connector socket.io API proposal

### ISSUE THREAD

**Issue by outSH** [hyperledger/cactus#1703]
We'd like to start a discussion on common socket.io API to communicate from `api-server` with any ledger-connector / validator using socket.io. This includes `ApiClient` design we all agree on, and documented socketio communication protocol to simplify adding new ledger-connectors / validators in the future.

@petermetz - please review and share your comments and opinions.

## SocketIOApiClient
- We want the connection protocols used by the `ledger-plugins` to be completely independent from each other. For instance, we want to separate axios interface (generated by OpenAPI) from socketio one, so that the `ledger-plugin` could implement just socketio interface without any use of openapi.
- So, in general, we want to have a separate `ApiClient` for socketio communication instead of single, common client as implemented in `BesuApiClient`.
- Reasons to separate:
    - This approach is in line with separation of concerns and open-close principles, it would simplify adding additional protocols / removing obsolete ones when needed (we can't make any assumptions regarding business needs of future users).
    - For validators with many protocols, it would be great to have opt in/out possibility for specific protocols in ledger-plugin during runtime, so that cactus setup with BLP that uses only socketio interface (or any other protocol) doesn't maintain excesive protocol endpoints. This'd be great improvement from operations and security point of view, because there's simply less stuff to care about.
    - It's easier to see which protocol is used underneath when developing and debuging the BLP.
- I think a separate `SocketIOApiClient` for each ledger is better than one, common for all, due to easier type-safety and the fact that each ledger can support different operations after all.
- The problem however with separate `SocketIOApiClient` for each ledger is massive code duplication. If we assume common socketio cactus protocol (see below), we could extract common operations to separate class and reuse them as mixins. Each ledger could also add own operations when needed.

### Pseudo-code example
- This is just an idea pitch, usage, names, etc.. will be improved later.

#### socketio-api.ts
``` typescript
/**
 * socket.io connection setup and management, all LedgerApi will extend this class and reuse connection it establishes.
 */
class SocketIOApi {}
```

#### socketio-api-mixing-monitor.ts
``` typescript
interface ISocketIOMonitor<T> {
    watchBlocks(): T;
}

/**
 * Mixin function that will extend ApiClient from parameter with watchBlocks()
 * watchBlocks will implement common logic for exchanging socketio messages with the ledger, will use already defined SocketIOApi from the base.
 */
function SocketIOMonitor<MonEvent, TBase extends new (...args: any[]) => SocketIOApi>(Base: TBase) {
  return class SocketIOMonitor extends Base implements ISocketIOMonitor<MonEvent> {
    watchBlocks(): MonEvent {
      // common watchBlock logic
    }
  };
}
```

#### some-ledger-api.ts
``` typescript
/**
 * Our implementation of ApiClient for given ledger
 * We can add some ledger-specific socketio methods here (if they are reused in couple of ledgers they can be extracted as another mixin)
 */
class LedgerApiBase extends SocketIOApi {
  // Operations unique to this ledger
  doSomeOp(arg1: any, arg2: any) {
    // Some logic
  }
}

/**
 * Block data format returned for some ledger
 */
type LedgerMonEvent = {
    id: number,
    nonce: number
}
/**
 * Extend our LedgerApi with common logic for watchBlock operations.
 * This can be repeated to implement all other operations supported by the ledger-plugin.
 */
const LedgerApi = SocketIOMonitor<LedgerMonEvent, typeof LedgerApiBase>(LedgerApiBase);
```

#### blp.ts
``` typescript
// Example usage

const api = new LedgerApi();
api.watchBlocks(); // returns LedgerMonEvent
api.doSomeOp(1, 2);

function someGenericFunc(api: ISocketIOMonitor<{id: number}>) {
    // Example how generic function can work with many ledger ApiClients.
    // This one will work with any ApiClient which ledger block data has an numeric id field.
    // Passing incompatible ApiClient cause immediate error.
    let block = api.watchBlocks();
    return block.id;
}
someGenericFunc(api);
```

## Cactus SocketIO Communication Protocol
 - As Peter noticed in PR #1660, our current socketio-only validators use different communication patterns than besu socketio.
- I think we should agree on a common socketio protocol for specific operations done on the ledger.
- Why:
    - Code reuse mentioned above (in `SocketIOApiClient`), faster development, easier to maintain.
    - Easier development. BLP implementation can use specific operations and know that they behave similarly between ledgers (like it's done in `cmd-socketio-server` right now).
- For now we're interested in operations that are already supported by `cmd-socketio-server`, in the future this could be extended to more granular ones: Operations we want to specify right now:
  - watchBlocks (start / stop monitoring)
  - sendSyncRequest
  - sendAsyncRequest
- Common protocol should fully support `api-server`. In particular, we need to support running different `ledger-plugin` on different endpoints on a single server (like hinted in besu-socketio).
- Should be documented to clarify `ledger-plugin` development in other languages (like our python one).
- This way our current socketio-api could be called `Legacy` then, and used until validators were updated to the common protocol.
- I can prepare protocol proposal later on, after we clarify the basics.

**Comment by petermetz** [comment]
@outSH Thank you for the detailed writeup. I'll put down some notes here for the sake of completeness:

> For instance, we want to separate axios interface (generated by OpenAPI) from socketio one, so that the ledger-plugin could implement just socketio interface without any use of openapi.

This was always possible in `cmd-api-server`, a plugin can choose to register any number of OpenAPI endpoints including zero. Same for the SocketIO endpoints and gRPC as well.

> I think a separate SocketIOApiClient for each ledger is better than one, common for all, due to easier type-safety and the fact that each ledger can support different operations after all.

I agree, if you look at the ledger plugins, they all have their own API client classes defined for the same reason.

> Common protocol should fully support api-server. In particular, we need to support running different ledger-plugin on different endpoints on a single server (like hinted in besu-socketio).

Yes, I just want to doubly highlight this as important. Not just that we can run multiple endpoints on a single server, but also that multiple endpoints can be used by a single client (such as a browser) in parallel on the same SocketIO connection via multiplexing otherwise if someone is building a web application that uses two different async endpoints they would need to maintain two separate SocketIO connections which is overhead.

**Comment by outSH** [comment]
@petermetz 

> Thank you for the detailed writeup. I'll put down some notes here for the sake of completeness (since the PR was already merged anyway)

This is not a description of PR #1660, this is a proposal for more complete change regarding socket.io interface, in particular common ApiClient design for all validators :) Please review it in free time and share any additional comments if possible (there's low priority on that, I think). This is general idea for now, I can provide more details later on.

> This was always possible in cmd-api-server, a plugin can choose to register any number of OpenAPI endpoints including zero. Same for the SocketIO endpoints and gRPC as well.

Good to know. Our main concern however, is that socketio interface is included in openapi.json specification, and the code later on derives from the api client generated by the openapi tool. This is a dependency we're concerned about, and we'd like to introduce a design where socketio endpoints would not be included in openapi specification at all :)

**Comment by petermetz** [comment]
> Our main concern however, is that socketio interface is included in openapi.json specification

@outSH OpenAPI does not have the ability to describe socketio endpoints. Do you mean the data model only or something else?

**Comment by outSH** [comment]
> OpenAPI does not have the ability to describe socketio endpoints. Do you mean the data model only or something else?

Yes, I mean that `ApiClient` derives from `DefaultApi`, and uses data model definitions generated by openapi (`"../generated/openapi/typescript-axios"`), hence forming unnecessary dependency. The exact types I have in mind in Besu connector are:

``` typescript
import {
  DefaultApi,
  WatchBlocksV1,
  WatchBlocksV1Progress,
} from "../generated/openapi/typescript-axios";
```

**Comment by petermetz** [comment]
@outSH I'm getting closer, thank you for the answers. Do you mean the `ApiClient` class in `packages/cactus-api-client/src/main/typescript/api-client.ts`? And if yes, by the imports, do you mean this line? `import { DefaultApi as ApiConsortium } from "@hyperledger/cactus-plugin-consortium-manual";` (or if not that one then which one?)

**Comment by outSH** [comment]
@petermetz No, I mean the ledger specific `ApiClient`, like the one for Besu - `packages/cactus-plugin-ledger-connector-besu/src/main/typescript/api-client/besu-api-client.ts`. I'd like to standarize implementation and communication protocol of those (for other ledgers) :)

**Comment by petermetz** [comment]
@outSH Gotcha. How would you represent it in the standardized implementation's data model (the request objects) when certain request parameters are mandatory for one ledger and optional (or non-existent) for another one? 
For example in the run transaction endpoint, besu has `consistencyStrategy` which does not have a close match in Corda from the notary's perspective (as far as I know, but do let me know if this is actually a bad example)

**Comment by outSH** [comment]
> How would you represent it in the standardized implementation's data model (the request objects) when certain request parameters are mandatory for one ledger and optional (or non-existent) for another one?

With generics, this is already described in the original post, please check `some-ledger-api.ts` pseudocode above :)

**Comment by petermetz** [comment]
> @petermetz
> 
> > How would you represent it in the standardized implementation's data model (the request objects) when certain request parameters are mandatory for one ledger and optional (or non-existent) for another one?
> 
> With generics, this is already described in the original post, please check `some-ledger-api.ts` pseudocode above :)

@outSH Sorry let me clarify: How would you represent it in the openapi.json files that power the code generation?

**Comment by outSH** [comment]
> Sorry let me clarify: How would you represent it in the openapi.json files that power the code generation?

@petermetz As mentioned in the proposal above ...

> For instance, we want to separate axios interface (generated by OpenAPI) from socketio one, so that the ledger-plugin could implement just socketio interface without any use of openapi.

... one of the major assumptions and goals is to remove dependency on `openapi.json` files and openapi infrastructure for socketio only connectors.

**Comment by petermetz** [comment]
> > Sorry let me clarify: How would you represent it in the openapi.json files that power the code generation?
> 
> @petermetz As mentioned in the proposal above ...
> 
> > For instance, we want to separate axios interface (generated by OpenAPI) from socketio one, so that the ledger-plugin could implement just socketio interface without any use of openapi.
> 
> ... one of the major assumptions and goals is to remove dependency on `openapi.json` files and openapi infrastructure for socketio only connectors.

@outSH Thank you, now I understand. Sorry for the very slow response! 

Alright, the bottom line is that you want to swim away from OpenAPI as much as possible and I'd prefer to swim towards it. 
We just need to figure out how to do that without making the code and the documentation confusing to users, e.g. I don't want to stop you from doing it, but I need to make sure that the rest of the code is not in the blast radius for the side-effects of this change. Do you have any suggestions for this? On a high level I'm thinking of adding more packages as needed to contain it that way. I created 2 issues that are loosely related to this as well:
https://github.com/hyperledger/cactus/issues/1886
https://github.com/hyperledger/cactus/issues/1885

**Comment by outSH** [comment]
@petermetz Well, my goal was to change all packages, including Accenture ones, so this is problematic :( Why do you want to make every connector strictly dependent on `openapi.json`? Why are we forcing this on connectors that doesn't use openapi at all? I thought the assumption was that all plugins are opinionated

**Comment by RafaelAPB** [comment]
@outSH is the concern enforcing a standard data model for all plugins?

**Comment by outSH** [comment]
Discussion not relevant anymore, closing
